designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the language level, `static_matrix` can replace a `matrix` type if subsets of the matrix are never assigned to. That makes code such as the below illegal.
+
+```stan
+static_matrix[N, M] A = // Fill in A...
+A[10, 10] = 5;
+```
+
+Besides that they can be thought of as standard matrix.
+
+At the Stan Math level a `static_matrix` is a `var_type<Eigen::Matrix<double, -1, -1>>` with an underlying pointer to a `vari_type<Eigen::Matrix<double, -1, -1>>`*. This means that accessors for the value and adjoint of `var_type` objects can access contiguous chunks of each part of the `var_type`. Any function that accept a `var_type<T>` will support static matrices.
```

Comment:
[optional] I would prefere `var_template` and `vari_template` to `var_type` and `vari_type`. All these are types anyway, but not all are templates.

Reply:
I have the same issue with `template`---everything's a template.  I'm not sure what to call this, though.  You probably need the design for `var_whatevs` in here too.

Reply:
Naming things, I'm gonna rattle off a whole bunch here and see what sticks

- `var_expr`
- `ad_type`
- `ad_pack`
- `var_value`
- `ad_value`
- `var_t`

Reply:
>everything's a template

Not really. `var` and `vari` as they are implemented in develop are not templates. Even with the proposed changes they would not be templates - they would be template instatinations. So I still think `var_template` is the best way to differentiate that from existing `var`.

Reply:
What about just `var_t`?

Reply:
We already use `_t` for metaprograms returning a type ...

Reply:
> Even with the proposed changes they would not be templates

Is there any downside to just making them templates?

Reply:
Oh I see: https://github.com/stan-dev/math/issues/1805#issuecomment-618227258